lib/linalg.dx

```diff
@@ -144,35 +186,37 @@ def lu {n} (a: n=>n=>Float) :
   --      lRef!i!j := (a.i.j - s) / (get uRef!j!j)
   --    for i:n. ()
 
-    lmat = lower_tri_mat lRef
-    umat = upper_tri_mat uRef
+    -- Helper functions to index into triangular matrices
+    -- from indices that themselves need to be cast or projected
+    -- to the right type.
+    def unsafe_lmat {m} (i:n) (j:m) [Ix m] : Ref _ Float =
+      lower_tri_mat lRef i $ unsafe_cast j
+    def unsafe_umat {m} (i:n) (j:m) [Ix m, Subset (i..) m] : Ref _ Float =
+      upper_tri_mat uRef i $ from_just (project (i..) j)
+
     for j:n.
       for i:(..j).
         i' = inject n i
         s = sum for k:(..i).
           -- TODO: get transitivity of Subset working so we can do a single injection
-          k'' = inject (..j) k
-          k' = inject n k''
-          ukj = get $ umat k' ((unsafe_nat_diff (ordinal j) (ordinal k))@_)
-          lik = get $ lmat i' (cast k)
+          k' = inject n (inject (..j) k)
+          ukj = get $ unsafe_umat k' j
```

Comment:
So, if I understand what's happening here, this unsafe indexing operation should be removable, for this reason:  We have in scope
```
j: n
i: (..j)
k: (..i)
```
We want, for safely indexing uRef,
```
k3: n
j3: (k3..)
```
With `transpose_lower_ix i k` we can get the intermediates
```
k2: (..j)
i2: (k2..)
```
Then, including a `Subset` instance like the one suggested above, we should be able to just write
```
k3 = inject n k2
i3 = inject (k3..) i2
```

How much of the rest of these `unsafe_umat` and `unsafe_lmat` do you think we can clean up by similar tactics?